### DISCUSSION TITLE
Is there a recommended editor for the fonts?

### DISCUSSION THREAD

**Discussion by MarkusThur** [bigtreetech/BIGTREETECH-TouchScreenFirmware#1957] [Q&A] [answered]
Thank you for your help.

**Comment by Allochtone** [comment]
I was about to ask the same question. I looked at what font designer could read .fon files and I found FontForge and Fony but I couldn't read any of the provided files somehow. So I thought of doing my own by converting a .ttf to a .fon without any success. It seems more complex than I originally thought.

**Reply by MarkusThur** [reply]
I came to nearly same with those two and some more editors...
All that read "Bitmap Windows fonts" or create them, say this is not a valid Windows .fon file. But I also find no information which format it could be or what the proper designer is.

What I didn't try till now is to view those in a hex editor and may get a idea how it works. Neither I tired to find the position in the sources, where the characters are read from that files, that also could give a good hint, where to find a "designer" programm for it, or how to create one. It can't be that difficult.

**Comment by bigtreetech** [comment, marked as answer]
[FontMaker.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/6615004/FontMaker.zip)
Hello, here is a dot matrix font tool, but the language is Chinese. Moreover, it can only generate dot matrix font files, and cannot preview .fon files

![image](https://user-images.githubusercontent.com/38851044/121155937-d6307f80-c87a-11eb-8828-b0be59389e1d.png)
The .fon used by TFT is just a simple dot matrix bitmap font, As shown in the figure, each point represents a bit in .fon. And the scanning direction of bitmap font is from top to bottom and left to right.

Bye the way, We are designing our own dot matrix bitmap font tools, which will be completed in July, and it will be convenient to customize fonts by then

**Reply by MarkusThur** [reply]
Thank you very much... 
Think in the end we can wait for the tool to be ready.

I just played a bit with it.. (results see below line stop reading here, if you are interested)

A .cst or textfile containing the caracters that shall be represented by our fonts in the correct order and a "hint" in a .md file in the fonts folder would be nice.

I tested vs the '8x16_byte_ascii.fon'. If this font would contain all characters as it would be genrated from the tool above, this file would have to have 2 KByte and not 1.48KByte. `127 chars x 16 byte = 2.032 Byte + 2 Bytes EOF`.
So it's to asume that this file does leave out "nonprintable"  characters and the UI system is aware of it.
Think I saw some code pieces doing that.

The "unicode" font with 4.5 MByte and 24x24 bits per character seems to contain 65.536 character representations, that can be adressed with 16bit / character. So the order of characters there most probably is as given by UTF-16. Sadly the tool above does not contain a preset for that.
Think I also saw some code that does the resolving of UTF-8 encoding to the UTF-16 order of characters.

So presets with the "correct" order for both types of fonts would be nice also

Again thank you very much.

------------ stop reading here, if you don't need to see how the bits work --------------------

The tool can output binary as well has "human readable" bit representations of a font. I just did it for the Numbers `0123456789` with 8x16 size.. This is convenient, as most Hex editors represent 16 bytes in a line and every character then occupies one line.

Human readable result:

```
/******************************************************************************
* Font Width:8
* Font Height:16
* Font Count:10
*******************************************************************************/
//UNICODE:30
0x00,0x00,0x00,0x78,0x84,0x84,0x84,0x84,0x84,0x84,0x84,0x84,0x78,0x00,0x00,0x00,

//UNICODE:31
0x00,0x00,0x00,0x10,0x30,0x50,0x10,0x10,0x10,0x10,0x10,0x10,0x10,0x00,0x00,0x00,

//UNICODE:32
0x00,0x00,0x00,0x78,0x84,0x04,0x04,0x04,0x08,0x10,0x20,0x40,0xFC,0x00,0x00,0x00,

//UNICODE:33
0x00,0x00,0x00,0x78,0x84,0x04,0x04,0x38,0x04,0x04,0x04,0x84,0x78,0x00,0x00,0x00,

//UNICODE:34
0x00,0x00,0x00,0x08,0x18,0x18,0x28,0x28,0x48,0x48,0xFC,0x08,0x08,0x00,0x00,0x00,

//UNICODE:35
0x00,0x00,0x00,0x7C,0x40,0x40,0x80,0xF8,0x84,0x04,0x04,0x84,0x78,0x00,0x00,0x00,

//UNICODE:36
0x00,0x00,0x00,0x78,0x84,0x80,0x80,0xB8,0xC4,0x84,0x84,0x84,0x78,0x00,0x00,0x00,

//UNICODE:37
0x00,0x00,0x00,0xFC,0x08,0x08,0x10,0x10,0x20,0x20,0x40,0x40,0x40,0x00,0x00,0x00,

//UNICODE:38
0x00,0x00,0x00,0x78,0x84,0x84,0x84,0x78,0x84,0x84,0x84,0x84,0x78,0x00,0x00,0x00,

//UNICODE:39
0x00,0x00,0x00,0x78,0x84,0x84,0x84,0x8C,0x74,0x04,0x04,0x88,0x70,0x00,0x00,0x00,

```

binary result:

```
00000000: 0000 0078 8484 8484 8484 8484 7800 0000  ...x........x...
00000010: 0000 0010 3050 1010 1010 1010 1000 0000  ....0P..........
00000020: 0000 0078 8404 0404 0810 2040 fc00 0000  ...x...... @....
00000030: 0000 0078 8404 0438 0404 0484 7800 0000  ...x...8....x...
00000040: 0000 0008 1818 2828 4848 fc08 0800 0000  ......((HH......
00000050: 0000 007c 4040 80f8 8404 0484 7800 0000  ...|@@......x...
00000060: 0000 0078 8480 80b8 c484 8484 7800 0000  ...x........x...
00000070: 0000 00fc 0808 1010 2020 4040 4000 0000  ........  @@@...
00000080: 0000 0078 8484 8478 8484 8484 7800 0000  ...x...x....x...
00000090: 0000 0078 8484 848c 7404 0488 7000 0000  ...x....t...p...
000000a0: 0d0a                                     ..

```

As I had some problems to get it, what 'scan direction top to bottom and, left to right' means.
It's like the direction, you read e.g. english.
Start at top left corner, read to the right and then the next row below it.

------ stop reading here, if you dont need to see the Unicode 31 decoded ------

```
--------  00
--------  00
--------  00
---x----  10
--xx----  30
-x-x----  50
---x----  10
---x----  10
---x----  10
---x----  10
---x----  10
---x----  10
---x----  10
--------  00
--------  00
--------  00
```

**Comment by MarkusThur** [comment]
This document gives a hint on how to do it properly:
https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/blob/master/readme/English/How%20to%20customize%20fonts%202019-11-11.pdf

Somewhere in utf8_decode.c the system becomes aware, where to find a specific character in the .fon file.. Vice versa it determines, where it has to be

**Reply by AndylgTom** [reply]
> This document gives a hint on how to do it properly: https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/blob/master/readme/English/How%20to%20customize%20fonts%202019-11-11.pdf
> 
> Somewhere in utf8_decode.c the system becomes aware, where to find a specific character in the .fon file.. Vice versa it determines, where it has to be

Please update the link, the file is not available.

**Comment by AndylgTom** [comment]
Translated the program into English
![image](https://user-images.githubusercontent.com/20997012/175806134-97f167cf-e72f-4015-b36f-72f7101854a5.png)
[FontMakerenu.zip](https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/files/8986245/FontMakerenu.zip)
